File: cpp/behavioral/mediator.cpp
/*!
* \brief   Mediator.Define an object that encapsulates how 
*       a set of objects interact.Mediator promotes loose 
*       coupling by keeping objects from referring to each 
*       other explicitly,and it lets you vary their interaction
*       independently.
*
*     ( Renter and Landlord communicate through HouseMediator. 
*       So Renter and Landlord don't need to meet each other. )
*
*        Person     <=>     Mediator     <=>    Person
*          |                   |                  |
*        Renter           HouseMediator        Landlord
*/

#include <iostream>
#include <string>

class Mediator;

class Person {
public:
  virtual void SetMediator(Mediator *mediator) {}
  virtual void SendMessage(std::string &message) {}
  virtual void GetMessage(std::string &message) {}
protected:
  Mediator *mediator_;
};

class Mediator {
public:
  virtual void Send(std::string &message, Person *person) {}
  virtual void SetRenter(Person *renter) {}
  virtual void SetLandlord(Person *landlord) {}
};

// Concrete Person A
class Renter :public Person {
public:
  void SetMediator(Mediator *mediator) {
    mediator_ = mediator; 
    mediator_->SetRenter(this);
  }
  void SendMessage(std::string &message) { 
    mediator_->Send(message, this);
  }
  void GetMessage(std::string &message) { 
    std::cout << "Renter gets the message from Landlord£º" << message;
  }
};

// Concrete Person B
class Landlord :public Person {
public:
  void SetMediator(Mediator *mediator) {
    mediator_ = mediator; 
    mediator_->SetLandlord(this);
  }
  void SendMessage(std::string &message) {
    mediator_->Send(message, this);
  }
  void GetMessage(std::string &message) {
    std::cout << "Landlord gets the message from Renter£º" << message;
  }
};

// Concrete Mediator.
class HouseMediator :public Mediator {
public:
  HouseMediator() :renter_(NULL), landlord_(NULL) {}
  void SetRenter(Person *renter) { 
    renter_ = renter; 
  }
  void SetLandlord(Person *landlord) {
    landlord_ = landlord; 
  }
  void Send(std::string &message, Person *person) {
    if (person == renter_)
      landlord_->GetMessage(message);
    else
      renter_->GetMessage(message);
  }
private:
  Person *renter_;
  Person *landlord_;
};

int main() {
  Mediator *mediator = new HouseMediator();
  Person *renter = new Renter();
  Person *landlord = new Landlord();

  renter->SetMediator(mediator);
  landlord->SetMediator(mediator);

  renter->SendMessage(std::string("I want to rent a house in ShenZhen.\n"));
  landlord->SendMessage(std::string("I have a house in ShenZhen.\n"));
  
  delete renter;
  delete landlord;
  delete mediator;

  return 0;
}

File: python/behavioral/mediator.py
#!/usr/bin/python
# -*- coding : utf-8 -*-

"""
brief:
  Mediator.
    Define an object that encapsulates how 
  a set of objects interact.Mediator promotes loose 
  coupling by keeping objects from referring to each 
  other explicitly,and it lets you vary their interaction
  independently.

  ( Renter and Landlord communicate through HouseMediator. 
    So Renter and Landlord don't need to meet each other. )
    
     Person     <=>     Mediator     <=>    Person
       |                   |                  |
     Renter           HouseMediator        Landlord
"""

# Mediator
class HouseMediator: 
  def __init__(self):
    self.renter = None
    self.landlord = None
    
  def set_renter(self, renter):
    self.renter = renter
    
  def set_landlord(self, landlord):
    self.landlord = landlord
    
  def send_message(self, message, sender):
    if (sender == self.renter and
        self.landlord != None):
      self.landlord.get_message(message)
    elif (sender == self.landlord and 
          self.renter != None):
      self.renter.get_message(message)
    else:
      print("No reciever.")
    
# Person
class Person:
  def __init__(self):
    self.mediator = None
    
  def send_message(self, message):
    if self.mediator != None:
      self.mediator.send_message(message, self)
    else:
      print("No mediator.")

# Person A.
class Landlord(Person):  
  def set_mediator(self, mediator):
    self.mediator = mediator
    self.mediator.set_landlord(self)
    
  def get_message(self, message):
    print("Landlord gets the message from Renter：%s" % (message))
  
# Person B.
class Renter(Person):  
  def set_mediator(self, mediator):
    self.mediator = mediator
    self.mediator.set_renter(self)

  def get_message(self, message):
    print("Renter gets the message from Landlord: %s" % (message))
    
  
if __name__ == '__main__':

  renter = Renter()
  landlord = Landlord()
  mediator = HouseMediator()
  
  renter.set_mediator(mediator);
  landlord.set_mediator(mediator);
  
  renter.send_message("I want to rent a house in ShenZhen.");
  landlord.send_message("I have a house in ShenZhen.");
  
